assignment-1/pkg/controller/controller.go

```diff
@@ -0,0 +1,206 @@
+	}
+	go wait.Until(c.worker, time.Second, ch)
+	<-ch
+	return nil
+}
+
+func (c *Controller) worker() {
+	for c.processItems() {
+	}
+}
+
+func (c *Controller) processItems() bool {
+	item, shutdown := c.SnapshotQueue.Get()
+	if shutdown {
+		return false
+	}
+	c.SnapshotQueue.Forget(item)
+	key, err := cache.MetaNamespaceKeyFunc(item)
+	if err != nil {
+		fmt.Println(err.Error())
+		return false
+	}
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		fmt.Println(err.Error())
+		return false
+	}
+	snapshot, err := c.SnapshotLister.Snapshots(namespace).Get(name)
+	if err != nil {
+		fmt.Println(err.Error())
+		return false
+	}
+	if snapshot.Spec.Operation == "backup" {
+		err := c.createBackup(snapshot)
+		if err != nil {
+			fmt.Println(err.Error())
+			return false
+		}
+	} else if snapshot.Spec.Operation == "restore" {
+		err := c.restoreBackup(snapshot)
+		if err != nil {
+			fmt.Println(err.Error())
+			return false
+		}
+	}
+	return true
+}
+
+func (c *Controller) createBackup(snapshot *v1alpha1.Snapshot) error {
+	// PVC doesn't exists, return error
+	pvc, err := c.KubernetesClientSet.CoreV1().PersistentVolumeClaims(
+		snapshot.Namespace).Get(context.Background(), snapshot.Spec.PvcName,
+		metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	// if volumesnapshot exists, don't create
+	_, err = c.ExternalSnapshotClient.SnapshotV1().VolumeSnapshots(snapshot.Namespace).Get(
+		context.Background(), GetSnapshotName(pvc), metav1.GetOptions{})
+	if err == nil {
+		return err
+	}
+	return c.createSnapshot(pvc, snapshot)
+}
+
+func (c *Controller) createSnapshot(pvc *corev1.PersistentVolumeClaim, snapshot *v1alpha1.Snapshot) error {
+	snapshotName := GetSnapshotName(pvc)
+	snapshotClassname := GetSnapshotClassname()
+	snapTemplate := &volumesnapshotv1.VolumeSnapshot{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: snapshotName,
+		},
+		Spec: volumesnapshotv1.VolumeSnapshotSpec{
+			VolumeSnapshotClassName: &snapshotClassname,
+			Source: volumesnapshotv1.VolumeSnapshotSource{
+				PersistentVolumeClaimName: &pvc.Name,
+			},
+		},
+	}
+	_, err := c.ExternalSnapshotClient.SnapshotV1().VolumeSnapshots(pvc.Namespace).Create(
+		context.Background(), snapTemplate, metav1.CreateOptions{})
+	if err != nil {
+		return err
+	}
+	err = c.updateStatus(snapshotName, pvc, snapshot)
+	if err != nil {
+		return err
+	}
+	fmt.Printf("snapshot %s created\n", snapshotName)
+	return nil
+}
+
+func (c *Controller) updateStatus(snapshotName string, pvc *corev1.PersistentVolumeClaim, snapshot *v1alpha1.Snapshot) error {
+	snapshotDeepcopy := snapshot.DeepCopy()
+	snapshotDeepcopy.Status.VolumeSnapshotName = snapshotName
+	snapshotDeepcopy.Status.Storage = pvc.Spec.Resources.Requests.Storage().String()
+	snapshotDeepcopy.Status.StorageClassname = *pvc.Spec.StorageClassName
+	_, err := c.SnapshotClientSet.ShahpratikrV1alpha1().Snapshots(pvc.Namespace).Update(
+		context.Background(), snapshotDeepcopy, metav1.UpdateOptions{})
```

Comment:
instead of creating context.Background(), derive it from parent as arg to the function